### PR TITLE
FIX-#3208: Update pandas to 1.3.0

### LINKED ISSUE #3208: Update pandas to 1.3.0

### PR DESCRIPTION
**Pull request by YarShev**
Signed-off-by: Igoshev, Yaroslav <yaroslav.igoshev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3208  
- [ ] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/backends/base/query_compiler.py [hunk 1]
```diff
@@ -1488,6 +1495,7 @@ def pivot_table(
         dropna,
         margins_name,
         observed,
+        sort,
```
Comment (dchigarev) [inline modin/backends/base/query_compiler.py:1498]: `sort` is a new parameter for `pivot_table` introduced in pandas 1.3.0, so we have to pass it to the backend as well
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pivot_table.html

File: modin/backends/base/query_compiler.py [hunk 2]
```diff
@@ -1035,7 +1035,14 @@ def insert(self, loc, column, value):
         Returns:
             A new QueryCompiler with new data inserted.
         """
-        return DataFrameDefault.register(pandas.DataFrame.insert, inplace=True)(
+
+        def inserter(df, loc, column, value):
+            if isinstance(value, pandas.DataFrame):
```
Comment (dchigarev) [inline modin/backends/base/query_compiler.py:1040]: Handling of the frame cases now happens at the front-end so we do not expect to meet DataFrame here unless it was originally Series at the front-end

File: modin/pandas/base.py [hunk 3]
```diff
@@ -2759,8 +2759,10 @@ def transform(self, func, axis=0, *args, **kwargs):
         kwargs["is_transform"] = True
         try:
             result = self.agg(func, axis=axis, *args, **kwargs)
-        except Exception:
-            raise ValueError("Transform function failed")
+        except TypeError:
```
Comment (dchigarev) [inline modin/pandas/base.py:2794]: reflecting the same [pandas changes](https://github.com/pandas-dev/pandas/pull/40288/files#diff-2257b34410aee27eb14e348b9545fef2e212ff93bd72af02d700ae8df43d97bbL230)

File: modin/pandas/test/dataframe/test_udf.py [hunk 4]
```diff
@@ -144,6 +153,11 @@ def apply_func(series, y):
     )
 
 
+@pytest.mark.xfail(
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_udf.py:156]: this bug became visible because of the [changes of handling DataFrames](https://github.com/pandas-dev/pandas/issues/42403) at the `DataFrame.insert`

File: modin/pandas/base.py [hunk 5]
```diff
@@ -580,11 +620,8 @@ def _aggregate(self, func, *args, **kwargs):
         # Dictionaries have complex behavior because they can be renamed here.
         elif func is None or isinstance(func, dict):
             return self._default_to_pandas("agg", func, *args, **kwargs)
-        elif is_list_like(func) or callable(func):
-            kwargs.pop("is_transform", None)
-            return self.apply(func, axis=_axis, args=args, **kwargs)
-        else:
-            raise TypeError("type {} is not callable".format(type(func)))
```
Comment (dchigarev) [inline modin/pandas/base.py:587]: this validation is done in the following `self.apply` function

File: modin/pandas/base.py [hunk 6]
```diff
@@ -795,19 +833,12 @@ def apply(
                 return result._query_compiler
             return result
         elif isinstance(func, dict):
-            if axis == 1:
-                raise TypeError(
-                    "(\"'dict' object is not callable\", "
-                    "'occurred at index {0}'".format(self.index[0])
-                )
```
Comment (dchigarev) [inline modin/pandas/base.py:802]: this seems incorrect, pandas do not raise such exceptions no more
```python
>>> pandas.DataFrame({"a": [1, 2, 3], "b": [3, 4, 5]}).apply({0: "sum"}, axis=1)
0    4
dtype: int64
```

File: modin/pandas/base.py [hunk 7]
```diff
@@ -783,6 +820,7 @@ def apply(
         args=(),
         **kwds,
     ):
+        self._validate_function(func, on_invalid=AssertionError)
```
Comment (dchigarev) [inline modin/pandas/base.py:823]: 1.3.0 brought a bit of inconsistency in error types, `apply` and `aggregate` now [raise an `AssertionError`](https://github.com/pandas-dev/pandas/pull/38867/files#diff-2257b34410aee27eb14e348b9545fef2e212ff93bd72af02d700ae8df43d97bbR350) in case of an invalid function, when for `transform` [it's still `TypeError/ValueError`](https://github.com/pandas-dev/pandas/pull/38867/files#diff-2257b34410aee27eb14e348b9545fef2e212ff93bd72af02d700ae8df43d97bbR350), that's why this commit introduces `_validate_function` which can raise different types of exceptions

File: modin/pandas/test/dataframe/test_udf.py [hunk 8]
```diff
@@ -104,10 +105,18 @@ def test_aggregate_error_checking():
     agg_func_values + agg_func_except_values,
     ids=agg_func_keys + agg_func_except_keys,
 )
-def test_apply_type_error(func):
-    modin_df = pd.DataFrame(test_data["int_data"])
-    with pytest.raises(TypeError):
-        modin_df.apply({"row": func}, axis=1)
+def test_apply_key_error(func):
+    if not (is_list_like(func) or callable(func) or isinstance(func, str)):
+        pytest.xfail(
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_udf.py:110]: I propose to not take this soooooo corner case into account, we barely can fix this

File: modin/pandas/test/dataframe/test_udf.py [hunk 9]
```diff
@@ -100,6 +100,13 @@ def test_aggregate_error_checking():
         modin_df.aggregate("NOT_EXISTS")
 
 
+@pytest.mark.xfail(
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_udf.py:103]: 1. This is expression doesn't skip this test for the `BaseOnPython` backend, so the front-end logic of `apply` is still being tested.
2. This bug became visible now because we changed the "target" exception here from `TypeError` to the `KeyError` and now the apply's buggy behavior does not match for the test

File: modin/pandas/base.py [hunk 10]
```diff
@@ -333,6 +334,45 @@ def _validate_other(
...
+    def _validate_function(self, func, on_invalid=TypeError):
+        """
+        Check the validity of the function which is intended to be applied to the frame.
+
+        Parameters
+        ----------
+        func : object
+        on_invalid : callable(str), default: TypeError
+            Function to call in case invalid `func` is met, `on_invalid` takes an error
+            message as a single argument. If `on_invalid` has an Exception type then it
```
Comment (YarShev) [inline modin/pandas/base.py:346]:
```suggestion
            message as a single argument. If `on_invalid` has an ``Exception`` type then it
```

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 11]
```diff
@@ -465,3 +473,55 @@ def _read_csv_check_support(
...
+            )
+        if on_bad_lines is not None:
+            if error_bad_lines is not None or warn_bad_lines is not None:
+                raise ValueError(
+                    "Both on_bad_lines and error_bad_lines/warn_bad_lines are set. "
+                    "Please only set on_bad_lines."
+                )
+
+        if on_bad_lines not in ["error", "warn", "skip", None]:
+            raise ValueError(f"Argument {on_bad_lines} is invalid for on_bad_lines")
```
Comment (YarShev) [inline modin/experimental/engines/omnisci_on_ray/io.py:527]:
```suggestion
            raise ValueError(f"Argument {on_bad_lines} is invalid for on_bad_lines.")
```

File: modin/pandas/base.py [hunk 12]
```diff
@@ -333,6 +334,45 @@ def _validate_other(
...
+            will arise. Raise ``TypeError`` by default.
+            **Note:** This parameter is a hack to concord with pandas error types.
+        """
+
+        def error_raiser(msg, exception=Exception):
+            raise exception(msg)
+
+        if issubclass(on_invalid, Exception):
+            on_invalid = functools.partial(error_raiser, exception=on_invalid)
```
Comment (YarShev) [inline modin/pandas/base.py:355]: @dchigarev , why do we need this?

File: modin/pandas/test/test_io.py [hunk 13]
```diff
@@ -725,23 +725,26 @@ def test_read_csv_quoting(
         )
 
     # Error Handling parameters tests
-    @pytest.mark.xfail(
+    @pytest.mark.skip(
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:728]: Why skip?

File: modin/pandas/base.py (hunk 12 above)
Comment (dchigarev) [reply modin/pandas/base.py:355]: 1. We can't just use exception type as a function to throw an exception, there has to be explicit `raise`, so if `on_invalid` is some exception we have to wrap it into a function that will raise it.
2. Since `on_invalid` interface is `callable(str)` but for `error_raiser` it's `callable(str, Exception)` we need to bind a proper exception type to match `error_raiser` to the `on_invalid` contract

Comment (YarShev) [reply modin/pandas/base.py:355]: I see, make sense.

File: modin/pandas/test/test_io.py (hunk 13 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:728]: These tests were marked as skipped because bad lines handling doesn't supported in the Modin `read_csv`. Previously tests were passing but it wasn't guaranteed and now we can manually skip only failed tests but, i think, it is easier to skip full tests block.

File: modin/pandas/test/test_series.py [hunk 14]
```diff
@@ -721,13 +721,17 @@ def test_aggregate_numeric_except(request, data, func):
 
 @pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
 def test_aggregate_error_checking(data):
-    modin_series, _ = create_test_series(data)  # noqa: F841
+    modin_series, pandas_series = create_test_series(data)
 
+    assert pandas_series.aggregate("ndim") == 1
     assert modin_series.aggregate("ndim") == 1
     with pytest.warns(UserWarning):
```
Comment (anmyachev) [inline modin/pandas/test/test_series.py:728]: Does this mean that `UserWarning` should be appeared for function call on `modin_series` and `pandas_series`?
What if the warning appears for only one of the two series objects?

Comment (YarShev) [reply modin/pandas/test/test_series.py:728]: The warning should be appeared for `modin_series` only.

Comment (anmyachev) [reply modin/pandas/test/test_series.py:728]: I can imagine a situation where a warning will appear only for `pandas_series` object and the check will pass, but this is not what we expect.

Comment (YarShev) [reply modin/pandas/test/test_series.py:728]: Can you show an example?

File: modin/pandas/test/test_groupby.py [hunk 15]
```diff
@@ -1594,8 +1604,8 @@ def test_not_str_by(by, as_index):
     )
 
     modin_groupby_equals_pandas(md_grp, pd_grp)
-    df_equals(md_grp.sum(), pd_grp.sum())
-    df_equals(md_grp.size(), pd_grp.size())
-    df_equals(md_grp.agg(lambda df: df.mean()), pd_grp.agg(lambda df: df.mean()))
-    df_equals(md_grp.dtypes, pd_grp.dtypes)
-    df_equals(md_grp.first(), pd_grp.first())
+    eval_general(md_grp, pd_grp, lambda grp: grp.sum())
```
Comment (dchigarev) [inline modin/pandas/test/test_groupby.py:1607]: some of the corner cases at this test are not supported by pandas since now and raise an exception, wrapped function execution to `eval_generate` to check exceptions concordance

File: modin/backends/pandas/query_compiler.py [hunk 16]
```diff
@@ -2722,7 +2722,7 @@ def groupby_agg_builder(df, by=None, drop=False, partition_idx=None):
                 )
                 internal_by_cols = internal_by_df.columns
 
-                external_by = by.columns.difference(internal_by)
+                external_by = by.columns.difference(internal_by).unique()
```
Comment (dchigarev) [inline modin/backends/pandas/query_compiler.py:2725]: Here we need a set of columns without duplications, however new version of pandas returns *all* of the elements that are not presented at the `difference` argument, including duplicated ones
```python
>>> pandas125.Index(["a", "a", "a"]).difference([])
Index(['a'], dtype='object')
>>> pandas130.Index(["a", "a", "a"]).difference([])
Index(['a', 'a', 'a'], dtype='object')
```
so calling explicit `unique()` here to get a non-duplicating set

File: modin/pandas/test/test_series.py (hunk 14 above)
Comment (dchigarev) [reply modin/pandas/test/test_series.py]: @anmyachev is right, it's better to check for warnings throwing explicitly for modin Series, changed this test at [the last commit](https://github.com/modin-project/modin/pull/3209/commits/de68a8acafcfbf226a910a7ce95a11c9d8ec4df1)

File: modin/pandas/groupby.py [hunk 17]
```diff
@@ -194,41 +194,48 @@ def ndim(self):
...
                 result = result.loc[~mask_nan_rows]
             return result
 
         if freq is None and axis == 1 and self._axis == 0:
-            result = _shift(periods, freq, axis, fill_value)
+            no_by_data = (
+                self._df.drop(columns=self._by.columns)
+                if isinstance(self._by, BaseQueryCompiler)
+                else self._df
+            )
```
Comment (dchigarev) [inline modin/pandas/groupby.py:228]: new pandas version does not allow to shift from/to "by" columns, so filtering the source frame from "by" columns before shifting
```python
>>> data = {
...     "col1": [0, 1, 2, 3],
...     "col2": [4, 5, np.NaN, 7],
...     "col3": [np.NaN, np.NaN, 12, 10],
...     "col4": [17, 13, 16, 15],
...     "col5": [-4, -5, -6, -7],
... }
>>> pandas125.DataFrame(data).groupby(["col1", "col4", "col5"]).shift(axis=1)
   col1  col2  col3  col4  col5
0   NaN     0   4.0   NaN    17
1   NaN     1   5.0   NaN    13
2   NaN     2   NaN  12.0    16
3   NaN     3   7.0  10.0    15
>>> pandas130.DataFrame(data).groupby(["col1", "col4", "col5"]).shift(axis=1)
   col2  col3
0   NaN   4.0
1   NaN   5.0
2   NaN   NaN
3   NaN   7.0
```

File: modin/pandas/test/test_series.py [hunk 18]
```diff
@@ -3652,29 +3652,17 @@ def test_str_contains(data, pat, case, na):
...
-            pat, repl, n=n, case=case, regex=False
-        )
-    except Exception as e:
-        with pytest.raises(type(e)):
-            modin_series.str.replace(pat, repl, n=n, case=case, regex=False)
-    else:
-        modin_result = modin_series.str.replace(pat, repl, n=n, case=case, regex=False)
-        df_equals(modin_result, pandas_result)
-
+    eval_general(
```
Comment (dchigarev) [inline modin/pandas/test/test_series.py:3655]: the same test, but wrapped in `eval_general`, that has more proper logic of checking arisen exceptions for equality

File: modin/pandas/test/dataframe/test_udf.py (hunk 4 above)
Comment (YarShev) [reply modin/pandas/test/dataframe/test_udf.py:163]: @dchigarev, will you look at it after this pandas update?

File: modin/pandas/test/test_series.py (hunk 14 above)
Comment (YarShev) [reply modin/pandas/test/test_series.py]: @dchigarev , okay but I would like to look at the example anyway.

File: modin/pandas/test/dataframe/test_udf.py (hunk 8 above)
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_udf.py:117]: it's front-end responsibility to validate and sanitize parameters before passing them to the backend. If we don't do that then invalid parameters could propagate to the engine, which may perform functions lazily and do not raise any exception until the execution will be explicitly triggered, obviously, we want to notify that something went wrong ASAP. 

Also, backends that aren't backed by pandas DataFrames will not raise the same with exceptions anyway

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 19]
```diff
@@ -94,7 +95,7 @@ def _read(cls, filepath_or_buffer, **kwargs):
 
         # Now we need to define parameters, which are common for all partitions. These
         # parameters can be `sniffed` from empty dataframes created further
-        if names is None:
+        if names in [lib.no_default, None]:
```
[the remark this answers is not in the document]
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:98]: I think we have to leave it as is. Exactly here we need to know whether `names` parameter passed or not, so `lib.no_default` should be checked as well. In other cases `lib.no_default` and `None` values are almost the same (pandas [replaces](https://github.com/pandas-dev/pandas/blob/1.3.x/pandas/io/parsers/readers.py#L1308) `lib.no_default` with None).

File: modin/pandas/test/test_groupby.py [hunk 20]
```diff
@@ -1134,11 +1134,21 @@ def eval_shift(modin_groupby, pandas_groupby):
...
-        lambda groupby: groupby.shift(axis=1, fill_value=777),
-    )
+    # pandas introduced a bug in 1.3.0 that breaks shift in the following case,
+    # modin performs correctly, so we can't compare the result with pandas for now.
+    # You can track the pandas bug here:
+    # https://github.com/pandas-dev/pandas/issues/42401
+    try:
+        if pandas_groupby.ndim == 2:
+            pandas_groupby.shift(axis=1, fill_value=777)
+    except TypeError:
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py:1144]: done

File: modin/pandas/groupby.py (hunk 17 above)
Comment (dchigarev) [reply modin/pandas/groupby.py]: has rewritten in a more general way: now it drops 'by' columns not only in the "`BaseQueryCompiler` case"

File: modin/pandas/test/dataframe/test_map_metadata.py [hunk 21]
```diff
@@ -980,7 +980,12 @@ def test_insert(data):
     )
 
     # Bad inserts
-    eval_insert(modin_df, pandas_df, col="Bad Column", value=lambda df: df)
+    with pytest.raises(ValueError):  # see pandas issue #42403
+        if len(modin_df) == len(modin_df.columns):
+            # Error will not arise if the frame is a square-shaped
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_map_metadata.py]: applied

File: modin/pandas/test/dataframe/test_udf.py (hunk 9 above)
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_udf.py:103]: ok

File: modin/pandas/test/dataframe/test_udf.py (hunk 4 above)
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_udf.py:163]: ok

File: modin/pandas/test/dataframe/test_udf.py (hunk 8 above)
Comment (YarShev) [reply modin/pandas/test/dataframe/test_udf.py:117]: Make sense

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/engines/base/io/text/csv_dispatcher.py:98]: Ok

**Comment by YarShev** [comment]
@devin-petersohn, since we plan to release Modin 0.10.1 today, I think, we should postpone this pandas update for the next release because #3221 and #3219 need to be resolved in the same release as well. Thoughs?

**Comment by dchigarev** [comment]
> we should postpone this pandas update for the next release because #3221 and #3219 need to be resolved in the same release as well

these issues exist independently from the pandas update, they both became visible because we changed some tests in this PR, not because of the new pandas version, I don't see reasons why these issues block this PR

**Comment by YarShev** [comment]
Ok, I see. In that case let's go through the changes and if these are good to be merged today, we will put this pandas update in 0.10.1.

File: modin/conftest.py [hunk 22]
```diff
@@ -127,16 +127,16 @@ def simulate_cloud(request):
     assert IsExperimental.get(), "Simulated cloud must be started in experimental mode"
 
     from modin.experimental.cloud import create_cluster, get_connection
-    import pandas._testing
-    import pandas._libs.testing as cyx_testing
+    import modin.pandas.test.utils
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/conftest.py:130]: yes

**Review by devin-petersohn — APPROVED** [review]
Looks good to me! Thanks @YarShev and @dchigarev!
